Updated Legacy module LIP to address recent protocol changes (#169)
LiskHQ/lips: proposals/lip-0050.md

## Specification

### Module name

The Legacy module has the name `MODULE_NAME_LEGACY` (defined in the table below).

### Notation and Constants

We define the following constants:

| Name                           | Type  | Value  | Description                                                                                                |
|--------------------------------|-------| -------|------------------------------------------------------------------------------------------------------------|
| `MODULE_NAME_LEGACY`           | string|"legacy"| The name of the Legacy module.                                                                            |
| `COMMAND_RECLAIM`              | string|"reclaim"| The name of the reclaim command.                                                                         |
| `COMMAND_REGISTER_KEYS`        | string| "registerKeys" | The name of the register keys command.                                                                   |

> | `EVENT_NAME_ACCOUNT_RECLAIM`   | string | "accountReclaim" | The name of the account reclaim event.                                                                  |

Reviewer: ```suggestion
| `EVENT_NAME_ACCOUNT_RECLAIMED`   | string | "accountReclaimed" | The name of the account reclaimed event.                                                                  |
```

For consistency?

Author: Renamed

> | `EVENT_NAME_ACCOUNT_RECLAIMED`   | string | "accountReclaim" | The name of the account reclaimed event.                                                                  |

Reviewer: ```suggestion
| `EVENT_NAME_ACCOUNT_RECLAIMED`   | string | "accountReclaimed" | The name of the account reclaimed event.                                                                  |
```

| `EVENT_NAME_KEYS_REGISTERED`   | string | "keysRegistered" | The name of the keys registered event.                                                                  |
| `SUBSTORE_PREFIX_LEGACY_ACCOUNTS` | bytes | 0x0000 | Substore prefix of the legacy accounts substore. This contains the addresses and balances of legacy accounts. |
| `TOKEN_ID_LSK`       | bytes | 0x0000000000000000  | Identifier of the LSK token.                                        |

> | `NUM_BYTES_ADDRESS`            | uint32| 20   | The length of an address in bytes.    |

Reviewer: I would follow the convention `LENGTH_ADDRESS`, same for legacy address

Author: Ok, changed.

| `EVENT_NAME_RECLAIM_FAILED`   | string | "reclaimFailed" | The name of the reclaim failed event.                                                                  |
| `EVENT_NAME_KEYS_REGISTERED`   | string | "keysRegistered" | The name of the keys registered event.                                                                  |
| `EVENT_NAME_KEY_REGISTRATION_FAILED`   | string | "KeyRegistrationFailed" | The name of the key registration failed event.                                                                  |
| `SUBSTORE_PREFIX_LEGACY_ACCOUNTS` | bytes | 0x0000 | Substore prefix of the legacy accounts substore. This contains the addresses and balances of legacy accounts. |
| `TOKEN_ID_LSK`       | bytes | 0x0000000000000000  | Identifier of the LSK token.                                        |
| `NUM_BYTES_ADDRESS`            | uint32| 20   | The length of an address in bytes.    |
| `NUM_BYTES_LEGACY_ADDRESS`     | uint32| 8    | The length of a legacy address in bytes |
| `LENGTH_BLS_KEY`               | uint32| 48   | The length of a BLS key in bytes  |
| `LENGTH_PROOF_OF_POSSESSION`   | uint32| 96   | The length of a proof of possession in bytes |

> | `LENGTH_GENERATOR_KEY`         | uint32| 32   | The length of a generator key in bytes |

Author: Uniformized to `LENGTH`.

#### Functions from Other Modules

Calling a function `fct` from another module (named `ModuleName`) is represented by `ModuleName.fct(required inputs)`.

### Legacy Module Store

#### Legacy Accounts Substore

This substore contains an array with the addresses and the balances of all legacy accounts for which no [reclaim transaction][lip-0018#ReclaimTransaction] was included.

##### Substore Prefix, Store Key, and Store Value

* The substore prefix is set to `SUBSTORE_PREFIX_LEGACY_ACCOUNTS`.

> * The store key is an byte array of length `NUM_BYTES_LEGACY_ADDRESS` representing the legacy address.

Reviewer: ```suggestion
* The store key is a byte array of length `NUM_BYTES_LEGACY_ADDRESS` representing the legacy address.
```

Author: Fixed.

#### AccountReclaimed

> This event has `name = EVENT_NAME_ACCOUNT_RECLAIM`.

Reviewer: ```suggestion
This event has `name = EVENT_NAME_ACCOUNT_RECLAIMED`.
```

Author: Renamed

    "properties": {
        "legacyAddress": {
            "dataType": "bytes",
            "length": NUM_BYTES_LEGACY_ADDRESS,
            "fieldNumber": 1
        },
        "address": {
            "dataType": "bytes",
            "length": NUM_BYTES_ADDRESS,
            "fieldNumber": 2
        },
        "amount": {
            "dataType": "uint64",
            "fieldNumber": 3
        }
    }
}
```

> #### ReclaimFailed

Reviewer 2: After reading the command logic below, it seems that this event is only emitted during verification. It could then be removed (as we are not supposed to emit during verification).

```python
def verify(trs: Transaction) -> None:
    legacyAddress = getLegacyAddress(trs.senderPublicKey)
    if legacyAccounts(legacyAddress) is empty:
        address = SHA-256(trs.senderPublicKey)[:NUM_BYTES_ADDRESS]
        emitPersistentEvent(
            module = MODULE_NAME_LEGACY,
            name = EVENT_NAME_RECLAIM_FAILED,
            data = {
                "legacyAddress": legacyAddress,
                "address": address,
                "reason": RECLAIM_FAILED_NO_LEGACY_ACCOUNT
            },
            topics = [legacyAddress, address]

>         )

Reviewer 2: I don't think we need emit events during verification(s). If verification fails, the command never goes in the block chain.

Author: Indeed, removed events from verification.

        raise Exception('Public key does not correspond to a reclaimable account.')

    if legacyAccounts(legacyAddress).balance != trs.params.amount:
        address = SHA-256(trs.senderPublicKey)[:NUM_BYTES_ADDRESS]
        emitPersistentEvent(
            module = MODULE_NAME_LEGACY,
            name = EVENT_NAME_RECLAIM_FAILED,
            data = {
                "legacyAddress": legacyAddress,
                "address": address,
                "reason": RECLAIM_FAILED_INVALID_AMOUNT
            },
            topics = [legacyAddress, address]

>         )

Reviewer 2: same, I don't think we should emit here.

Author: Indeed, removed events from verification.

```python
def verify(trs: Transaction) -> None:
    validatorAddress = SHA-256(trs.senderPublicKey)[:LENGTH_ADDRESS]
    # an exception would also be raised if no validator account for this address
    if Validators.getValidatorAccount(validatorAddress).blsKey != INVALID_BLS_KEY:

>         raise Exception('Validator already has a registered keys.')

Reviewer 3: ```suggestion
        raise Exception('Validator already has a registered key.')
```

    legacyAddress = getLegacyAddress(trs.senderPublicKey)
    delete legacyAccounts(legacyAddress) from the legacy accounts substore

    newAddress = SHA-256(trs.senderPublicKey)[:NUM_BYTES_ADDRESS]
    # Unlock the tokens in the legacy reserve account and transfer them to the
    # account reclaiming the tokens.

>     token.unlock(ADDRESS_LEGACY_RESERVE, MODULE_NAME_LEGACY, TOKEN_ID_LSK, trs.params.amount)

Reviewer 2: I think that new convention says that we capitalize the module name.
```suggestion
    Token.unlock(ADDRESS_LEGACY_RESERVE, MODULE_NAME_LEGACY, TOKEN_ID_LSK, trs.params.amount)
```

Also below

Author: Changed everywhere.

```python
def verify(trs: Transaction) -> None:

    validatorAddress = 20-byte address derived from trs.senderPublicKey
    validatorAddress = SHA-256(trs.senderPublicKey)[:NUM_BYTES_ADDRESS]

>     if validator.getValidatorAccount(validatorAddress) does not exists :

Reviewer 2: I realize that here, we might not even need this condition, just calling the API will do. If the validator address does not exist, the function will throw an error (and try to emit, but that will do nothing here).

Author: Ah, a very good point! Removed the check.

        emitPersistentEvent(
            module = MODULE_NAME_LEGACY,
            name = EVENT_NAME_KEY_REGISTRATION_FAILED,
            data = {
                "address": validatorAddress,
                "reason": KEY_REGISTRATION_FAILED_NO_VALIDATOR_ACCOUNT
            },
            topics = [validatorAddress]

>         )

Reviewer 2: Same as for reclaim, no emitting in the verify function.

Author: Indeed, removed events from verification.

        raise Exception('Public key does not correspond to a registered validator.')

    if Validators.getValidatorAccount(validatorAddress).blsKey != INVALID_BLS_KEY:

>         emitPersistentEvent(
>             module = MODULE_NAME_LEGACY,
>             name = EVENT_NAME_KEY_REGISTRATION_FAILED,
>             data = {
>                 "address": validatorAddress,
>                 "reason": KEY_REGISTRATION_FAILED_KEYS_REGISTERED
>             },
>             topics = [validatorAddress]
>         )

Reviewer 2: ```suggestion
```
